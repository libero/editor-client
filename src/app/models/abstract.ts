import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as abstractConfig from 'app/models/config/abstract.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter abstract')]
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter impact statement')]
  });
}
