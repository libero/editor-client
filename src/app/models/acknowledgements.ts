import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { Step } from 'prosemirror-transform';
import { gapCursor } from 'prosemirror-gapcursor';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from './utils';
import { PlaceholderPlugin } from './plugins/placeholder.plugin';
import { SelectionPlugin } from './plugins/selection.plugins';

export function createAcknowledgementsState(content?: Element, changeSteps?: [Step]): EditorState {
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter acknowledgements')]
  });
}
