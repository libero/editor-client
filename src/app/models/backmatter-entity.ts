import { v4 as uuidv4 } from 'uuid';

import { JSONObject } from '../types/utility.types';

export abstract class BackmatterEntity {
  constructor() {
    this._id = uuidv4();
  }

  get id(): string {
    return this._id;
  }

  protected _id: string;

  protected abstract fromXML(xmlNode: Element): void;
  protected abstract fromJSON(json: JSONObject): void;
  protected abstract createBlank(): void;

  protected createEntity(data?: Element | JSONObject): void {
    if (!data) {
      this.createBlank();
    } else if (data instanceof Element) {
      this.fromXML(data);
    } else {
      this.fromJSON(data);
    }
  }
}
