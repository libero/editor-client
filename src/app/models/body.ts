import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
