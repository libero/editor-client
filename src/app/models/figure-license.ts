import { getTextContentFromPath } from 'app/models/utils';

export interface FigureLicense {
  copyrightHolder: string;
  copyrightStatement: string;
  copyrightYear: string;
  licenseType: string;
}

export const FIGURE_LICENSE_CC0 = 'LICENSE_CCO';
export const FIGURE_LICENSE_CC_BY_4 = 'LICENSE_CC_BY_4';
export const FIGURE_LICENSE_OTHER = 'LICENSE_OTHER';

export const FIGURE_LICENSE_SELECT_OPTIONS = [
  { label: 'CCO', value: FIGURE_LICENSE_CC0 },
  { label: 'CC_BY_4', value: FIGURE_LICENSE_CC_BY_4 },
  { label: 'Other', value: FIGURE_LICENSE_CC_BY_4 }
];

const LICENSE_URLS_MAP = {
  'http://creativecommons.org/licenses/by/4.0/': FIGURE_LICENSE_CC_BY_4,
  'http://creativecommons.org/publicdomain/zero/1.0/': FIGURE_LICENSE_CC0
};

export function createFigureLicenseAttributes(el: Element): FigureLicense {
  return {
    copyrightHolder: getTextContentFromPath(el, 'copyright-holder'),
    copyrightStatement: getTextContentFromPath(el, 'copyright-statement'),
    copyrightYear: getTextContentFromPath(el, 'copyright-year'),
    licenseType: getLicenseType(el)
  };
}

export function createEmptyLicenseAttributes(): FigureLicense {
  return {
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: undefined
  };
}

function getLicenseType(el: Element): string {
  const licenseUrl =
    getTextContentFromPath(el, 'license_ref') || el.querySelector('license').getAttribute('xlink:href') || '';

  return LICENSE_URLS_MAP[licenseUrl] || FIGURE_LICENSE_OTHER;
}
