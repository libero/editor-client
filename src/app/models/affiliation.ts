import { getTextContentFromPath } from 'app/models/utils';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Affiliation extends BackmatterEntity {
  label: string;
  institution: { name: string };
  address: { city: string };
  country: string;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  public getDisplayName(): string {
    return [this.institution.name, this.address.city, this.country].filter(Boolean).join(', ');
  }

  public clone(): Affiliation {
    const affitiation = new Affiliation();
    affitiation._id = this._id;
    affitiation.label = this.label;
    affitiation.country = this.country;
    affitiation.institution = { name: this.institution.name };
    affitiation.address = { city: this.address.city };
    return affitiation;
  }

  protected fromXML(xml: Element): void {
    this._id = xml.getAttribute('id');
    this.label = getTextContentFromPath(xml, 'label');
    this.institution = {
      name: [
        getTextContentFromPath(xml, 'institution[content-type="dept"]'),
        getTextContentFromPath(xml, 'institution:not([content-type])')
      ]
        .filter(Boolean)
        .join(', ')
    };

    this.address = {
      city: getTextContentFromPath(xml, 'addr-line named-content[content-type="city"]')
    };
    this.country = getTextContentFromPath(xml, 'country');
  }

  protected fromJSON(json: JSONObject): void {
    this._id = json.id as string;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }
}

export function createAffiliationsState(affiliationsXml: Element[]): Affiliation[] {
  return affiliationsXml.map((xml) => new Affiliation(xml));
}
