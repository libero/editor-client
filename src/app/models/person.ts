import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';
import { history } from 'prosemirror-history';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';

import * as bioConfig from 'app/models/config/author-bio.config';
import { Affiliation } from 'app/models/affiliation';
import { getTextContentFromPath, makeSchemaFromConfig } from 'app/models/utils';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Person extends BackmatterEntity {
  firstName: string;
  lastName: string;
  suffix?: string;
  isAuthenticated?: boolean;
  orcid?: string;
  email?: string;
  bio?: EditorState;
  isCorrespondingAuthor?: boolean;
  affiliations?: string[];
  hasCompetingInterest?: boolean;
  competingInterestStatement?: string;

  constructor(data?: JSONObject | Element, notesXml?: Element) {
    super();
    this.createEntity(data);
    if (data instanceof Element && notesXml instanceof Element) {
      this.setCompetingInterests(data, notesXml);
    }
  }

  public getDisplayName(): string {
    return [this.firstName, this.lastName, this.suffix].filter((_) => _).join(' ');
  }

  public getAffiliationsLabels(affiliations: Affiliation[]): string[] {
    return this.affiliations
      .map((affiliationId) => {
        const affiliation = affiliations.find(({ id }) => id === affiliationId);
        return affiliation ? affiliation.label : undefined;
      })
      .filter(Boolean)
      .sort((a, b) => Number(a) - Number(b));
  }

  clone(): Person {
    const person = new Person();
    person.firstName = this.firstName;
    person.lastName = this.lastName;
    person.suffix = this.suffix;
    person.isAuthenticated = this.isAuthenticated;
    person.email = this.email;
    person.bio = this.bio;
    person.isCorrespondingAuthor = this.isCorrespondingAuthor;
    person.affiliations = [...this.affiliations];
    person.hasCompetingInterest = this.hasCompetingInterest;
    person.competingInterestStatement = this.competingInterestStatement;

    return person;
  }

  protected fromXML(xml: Element): void {
    const orcidEl = xml.querySelector('contrib-id[contrib-id-type="orcid"]');

    this._id = xml.getAttribute('id') || this._id;
    this.firstName = getTextContentFromPath(xml, 'name > given-names');
    this.lastName = getTextContentFromPath(xml, 'name > surname');
    this.suffix = getTextContentFromPath(xml, 'name > suffix');
    this.isAuthenticated = orcidEl ? orcidEl.getAttribute('authenticated') === 'true' : false;
    this.orcid = orcidEl ? this.getOrcid(orcidEl.textContent) : '';
    this.bio = this.createBioEditorStateFromXml(xml.querySelector('bio'));
    this.email = getTextContentFromPath(xml, 'email');
    this.isCorrespondingAuthor = xml.getAttribute('corresp') === 'yes';
    this.affiliations = Array.from(xml.querySelectorAll('xref[ref-type="aff"]')).map((xRef) =>
      xRef.getAttribute('rid')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this._id = (json.id as string) || this._id;
    this.firstName = json.firstName as string;
    this.lastName = json.lastName as string;
    this.suffix = json.suffix as string;
    this.isAuthenticated = json.isAuthenticated as boolean;
    this.orcid = json.orcid as string;
    this.bio = json.bio
      ? this.createBioEditorStateFromJSON(json.bio as JSONObject)
      : this.createBioEditorStateFromXml();
    this.email = json.email as string;
    this.isCorrespondingAuthor = json.isCorrespondingAuthor as boolean;
    this.affiliations = (json.affiliations as string[]) || [];
  }

  protected createBlank(): void {
    this.firstName = '';
    this.lastName = '';
    this.suffix = '';
    this.isAuthenticated = false;
    this.orcid = '';
    this.bio = this.createBioEditorStateFromXml();
    this.email = '';
    this.isCorrespondingAuthor = false;
    this.affiliations = [];
  }

  private getOrcid(orcidUrl: string): string {
    const matches = orcidUrl.match(/(([0-9]{4}-?){4})/g);
    if (matches) {
      return matches[0];
    }
    return '';
  }

  private setCompetingInterests(dataXml: Element, notesXml: Element): void {
    const competingInterestEl = Array.from(notesXml.querySelectorAll('[fn-type="COI-statement"]')).find(
      (fnEl: Element) => {
        const id = fnEl.getAttribute('id');
        return dataXml.querySelector(`xref[ref-type="fn"][rid="${id}"]`);
      }
    );

    this.hasCompetingInterest = competingInterestEl
      ? competingInterestEl.textContent !== 'No competing interests declared'
      : false;
    this.competingInterestStatement = competingInterestEl ? competingInterestEl.textContent : '';
  }

  private createBioEditorStateFromXml(bio?: Element): EditorState {
    const schema = makeSchemaFromConfig(bioConfig.topNode, bioConfig.nodes, bioConfig.marks);
    return EditorState.create({
      doc: bio ? ProseMirrorDOMParser.fromSchema(schema).parse(bio) : undefined,
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), history()]
    });
  }

  private createBioEditorStateFromJSON(json: JSONObject): EditorState {
    const blankState = this.createBioEditorStateFromXml();
    return EditorState.fromJSON(
      {
        schema: blankState.schema,
        plugins: blankState.plugins
      },
      json
    );
  }
}

export function createAuthorsState(authorsXml: Element[], notesXml: Element | undefined): Person[] {
  return authorsXml.map((xml) => new Person(xml, notesXml));
}
