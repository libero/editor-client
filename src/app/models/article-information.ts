import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { buildInputRules } from './plugins/input-rules';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';

import { getTextContentFromPath, makeSchemaFromConfig } from './utils';
import * as licenseTextConfig from './config/license-text.config';
import { Person } from './person';
import moment from 'moment';
import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from '../types/utility.types';

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

const LICENSE_CC_BY_4_TEXT = `This article is distributed under the terms of the 
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/licenses/by/4.0/">
        Creative Commons Attribution License</ext-link>, which permits unrestricted use and redistribution provided that the original author 
    and source are credited.`;

const LICENSE_CC0_TEXT = `This is an open-access article, free of all copyright, and may be freely reproduced, 
    distributed, transmitted, modified, built upon, or otherwise used by anyone for any lawful purpose.
    The work is made available under the
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/publicdomain/zero/1.0/">Creative Commons CC0 public domain dedication</ext-link>.`;

export class ArticleInformation extends BackmatterEntity {
  public static getLicenseText(licenseType: string): EditorState {
    const paragraph = document.createElement('p');
    paragraph.innerHTML = {
      [LICENSE_CC_BY_4]: LICENSE_CC_BY_4_TEXT,
      [LICENSE_CC0]: LICENSE_CC0_TEXT
    }[licenseType];

    return ArticleInformation.createLicenseEditorState(paragraph);
  }

  public static createLicenseEditorState(content?: Element): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = document.implementation.createDocument('', '', null);

    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor()]
    });
  }

  articleType: string;
  dtd: string;
  articleDOI: string;
  publisherId: string;
  volume: string;
  elocationId: string;
  subjects: Array<string>;
  publicationDate: string;
  licenseType: string;
  copyrightStatement: string;
  licenseText: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (data instanceof Element && authors) {
      this.updateCopyrightStatementFromXml(data, authors);
    }
  }

  public clone(): ArticleInformation {
    const newArticleInformation = new ArticleInformation();
    newArticleInformation.articleType = this.articleType;
    newArticleInformation.dtd = this.dtd;
    newArticleInformation.articleDOI = this.articleDOI;
    newArticleInformation.elocationId = this.elocationId;
    newArticleInformation.volume = this.volume;
    newArticleInformation.publisherId = this.publisherId;
    newArticleInformation.subjects = [...this.subjects];
    newArticleInformation.copyrightStatement = this.copyrightStatement;
    newArticleInformation.licenseType = this.licenseType;
    newArticleInformation.publicationDate = this.publicationDate;
    newArticleInformation.licenseText = this.licenseText;

    return newArticleInformation;
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
          ? `${authors[0].lastName} et al`
          : '';
      const date = moment(this.publicationDate);

      this.copyrightStatement = `Â© ${date.isValid() ? date.format('YYYY') : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="major-subject"] subject')).map(
      (el: Element) => el.textContent
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const emptyEditorState = ArticleInformation.createLicenseEditorState();
    this.licenseText = EditorState.fromJSON(
      {
        schema: emptyEditorState.schema,
        plugins: emptyEditorState.plugins
      },
      json.licenseText as JSONObject
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        pubDateNode.querySelector('year')?.textContent,
        pubDateNode.querySelector('month')?.textContent,
        pubDateNode.querySelector('day')?.textContent
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
