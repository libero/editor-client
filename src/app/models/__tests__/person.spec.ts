import { createAuthorsState, Person } from 'app/models/person';

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

describe('Person model helpers', () => {
  it('creates an author with specified data', () => {
    const xmlData = {
      firstName: 'Test',
      lastName: 'Testerson'
    };

    expect(new Person(xmlData)).toMatchSnapshot();
  });

  it('creates an author with specified data and ID', () => {
    const xmlData = {
      firstName: 'Test',
      lastName: 'Testerson'
    };

    expect(new Person(xmlData)).toMatchSnapshot();
  });

  it('produces author display name with first name last name and suffix', () => {
    const author = {
      id: 'SUPER_UNIQUE',
      firstName: 'Test',
      lastName: 'Testerson',
      suffix: 'Capt.'
    };

    expect(new Person(author).getDisplayName()).toEqual('Test Testerson Capt.');
  });

  it('produces author display even when some details are missing', () => {
    const author = {
      id: 'SUPER_UNIQUE',
      firstName: 'Total',
      lastName: undefined,
      suffix: 'Commander'
    };

    expect(new Person(author).getDisplayName()).toEqual('Total Commander');
  });

  it('creates authors state', () => {
    const authorsContainer = document.createElement('div');
    authorsContainer.innerHTML = `
        <name><surname>Atherden</surname><given-names>Fred</given-names><suffix>Capt.</suffix></name>
        <contrib-id authenticated="true" contrib-id-type="orcid">https://orcid.org/0000-0002-6048-1470</contrib-id>
        <email>f.atherden@elifesciences.org</email>`;

    expect(createAuthorsState([authorsContainer], document.createElement('author-notes'))).toMatchSnapshot();
  });
});
