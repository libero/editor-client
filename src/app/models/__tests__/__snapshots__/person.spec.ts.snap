// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person model helpers creates an author with specified data 1`] = `
Person {
  "_id": "unique_id",
  "affiliations": Array [],
  "bio": Object {
    "doc": Object {
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
    "selection": Object {
      "anchor": 1,
      "head": 1,
      "type": "text",
    },
  },
  "competingInterestStatement": undefined,
  "email": undefined,
  "firstName": "Test",
  "hasCompetingInterest": undefined,
  "isAuthenticated": undefined,
  "isCorrespondingAuthor": undefined,
  "lastName": "Testerson",
  "orcid": undefined,
  "suffix": undefined,
}
`;

exports[`Person model helpers creates an author with specified data and ID 1`] = `
Person {
  "_id": "unique_id",
  "affiliations": Array [],
  "bio": Object {
    "doc": Object {
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "doc",
    },
    "selection": Object {
      "anchor": 1,
      "head": 1,
      "type": "text",
    },
  },
  "competingInterestStatement": undefined,
  "email": undefined,
  "firstName": "Test",
  "hasCompetingInterest": undefined,
  "isAuthenticated": undefined,
  "isCorrespondingAuthor": undefined,
  "lastName": "Testerson",
  "orcid": undefined,
  "suffix": undefined,
}
`;

exports[`Person model helpers creates authors state 1`] = `
Array [
  Person {
    "_id": "unique_id",
    "affiliations": Array [],
    "bio": Object {
      "doc": Object {
        "content": Array [
          Object {
            "type": "paragraph",
          },
        ],
        "type": "doc",
      },
      "selection": Object {
        "anchor": 1,
        "head": 1,
        "type": "text",
      },
    },
    "competingInterestStatement": "",
    "email": "f.atherden@elifesciences.org",
    "firstName": "Fred",
    "hasCompetingInterest": false,
    "isAuthenticated": true,
    "isCorrespondingAuthor": false,
    "lastName": "Atherden",
    "orcid": "0000-0002-6048-1470",
    "suffix": "Capt.",
  },
]
`;
